File: darkwater-core/src/main/java/kiinse/plugins/darkwaterapi/core/files/statistic/Statistic.java
package kiinse.plugins.darkwaterapi.core.files.statistic;

import kiinse.plugins.darkwaterapi.api.files.statistic.DarkStatistic;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

public class Statistic implements DarkStatistic {

    private final UUID player;
    private Map<EntityType, Integer> stats = new HashMap<>();

    protected Statistic(@NotNull UUID player, @NotNull JSONObject json) {
        this.player = player;
        load(json);
    }

    protected Statistic(@NotNull Player player, @NotNull JSONObject json) {
        this.player = player.getUniqueId();
        load(json);
    }

    private void load(@NotNull JSONObject jsonObject) {
        for (var key : jsonObject.keySet()) {
            stats.put(EntityType.valueOf(key), jsonObject.getInt(key));
        }
    }

    @Override
    public @NotNull UUID getPlayerUUID() {
        return player;
    }

    @Override
    public @NotNull Map<EntityType, Integer> getAllStatistic() {
        return stats;
    }

    @Override
    public int getStatistic(@NotNull EntityType type) {
        if (stats.containsKey(type)) {
            return stats.get(type);
        }
        return 0;
    }

    @Override
    public @NotNull DarkStatistic setStatistic(@NotNull EntityType type, int amount) {
        stats.remove(type);
        stats.put(type, amount);
        return this;
    }

    @Override
    public @NotNull DarkStatistic setStatistic(@NotNull Map<EntityType, Integer> stats) {
        this.stats = stats;
        return this;
    }

    @Override
    public @NotNull DarkStatistic addStatistic(@NotNull EntityType type) {
        stats.put(type, getStatistic(type) + 1);
        return this;
    }

    @Override
    public @NotNull JSONObject toJSONObject() {
        var json = new JSONObject();
        for (var stat : stats.entrySet()) {
            json.put(stat.getKey().toString(), stat.getValue());
        }
        return json;
    }
}

File: core/src/main/kotlin/kiinse/me/plugins/darkwaterapi/core/files/statistic/Statistic.kt
package kiinse.me.plugins.darkwaterapi.core.files.statistic

import kiinse.me.plugins.darkwaterapi.api.files.statistic.DarkStatistic
import org.bukkit.entity.EntityType
import org.bukkit.entity.Player
import org.json.JSONObject
import java.util.*

open class Statistic : DarkStatistic {

    final override val playerUUID: UUID
    private var stats: MutableMap<EntityType, Int> = EnumMap(org.bukkit.entity.EntityType::class.java)

    protected constructor(player: UUID, json: JSONObject) {
        playerUUID = player
        load(json)
    }

    protected constructor(player: Player, json: JSONObject) {
        playerUUID = player.uniqueId
        load(json)
    }

    private fun load(jsonObject: JSONObject) {
        for (key in jsonObject.keySet()) {
            stats[EntityType.valueOf(key)] = jsonObject.getInt(key)
        }
    }

    override val allStatistic: Map<EntityType, Int>
        get() = stats

    override fun getStatistic(type: EntityType): Int {
        return if (stats.containsKey(type)) {
            stats[type]!!
        } else 0
    }

    override fun setStatistic(type: EntityType, amount: Int): DarkStatistic {
        stats.remove(type)
        stats[type] = amount
        return this
    }

    override fun setStatistic(stats: MutableMap<EntityType, Int>): DarkStatistic {
        this.stats = stats
        return this
    }

    override fun addStatistic(type: EntityType): DarkStatistic {
        stats[type] = getStatistic(type) + 1
        return this
    }

    override fun toJSONObject(): JSONObject {
        val json = JSONObject()
        for ((key, value) in stats) {
            json.put(key.toString(), value)
        }
        return json
    }
}